Return ICompletableFuture from async methods in IMap API

vbekiaris opened the pull request:
`IMap` methods `getAsync`, `putAsync`, `setAsync`, `removeAsync` now return `ICompletableFuture` allowing for execution of callbacks. A new interface `BaseAsyncMap` has been introduced to maintain binary backwards compatibility, so code compiled against `IMap` API definition in version 3.6.x should continue working in 3.7.
Fixes #6999. Binary compatibility was tested with a sample client compiled against 3.6 as well as sigtest(https://wiki.openjdk.java.net/display/CodeTools/sigtest) tool which reports `setAsync` method was added to IMap interface (3.7-SNAPSHOT versus 3.6) but no other errors.

jerrinot commented on hazelcast/src/main/java/com/hazelcast/core/IMap.java:
```diff
@@ -299,7 +299,7 @@
      * @see java.util.concurrent.Future
      */
-    Future<V> getAsync(K key);
+    ICompletableFuture<V> getAsync(K key);
```
shouldn't we change the examples in JavaDoc?

jerrinot commented on hazelcast/src/main/java/com/hazelcast/core/BaseAsyncMap.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Base interface for a {@link BaseMap} providing additional {@link Future}-returning
```
could you describe why it in important to keep this?

jerrinot commented on hazelcast/src/main/java/com/hazelcast/core/BaseAsyncMap.java:
```diff
@@ -0,0 +1,39 @@
+ * Base interface for a {@link BaseMap} providing additional {@link Future}-returning
+ * *Async methods.
+ */
+public interface BaseAsyncMap<K, V> extends BaseMap<K, V> {
```
maybe a name indicating the intent would be better. `LegacyAsyncMap` or something.

jerrinot commented on hazelcast/src/main/java/com/hazelcast/core/BaseAsyncMap.java:
```diff
@@ -0,0 +1,39 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.core;
```
it belong to impl/internal package

vbekiaris commented:
Base interface was renamed to `LegacyAsyncMap` and moved to `com.hazelcast.map.impl` package. Amended javadoc in `IMap` to describe usage of `ICompletableFuture` return type. Updated `IMap.submitToKey` to also return `ICompletableFuture` (was returning a `j.u.c.Future` previously).

jerrinot commented:
:+1:

tombujok commented:
:+1: (just a note that I don't like our ICompletableFuture - should be changed in 4.0 IMHO)
